packages/notifications/push/amplify_push_notifications/lib/src/amplify_push_notifications_impl.dart

```diff
@@ -0,0 +1,394 @@
+    if (!_isConfigured) {
+      throw _needsConfigurationException;
+    }
+    final result = await _hostApi.getPermissionStatus();
+    switch (result.status) {
+      case PermissionStatus.denied:
+        return PushNotificationPermissionStatus.denied;
+      case PermissionStatus.granted:
+        return PushNotificationPermissionStatus.granted;
+      case PermissionStatus.shouldExplainThenRequest:
+        return PushNotificationPermissionStatus.shouldExplainThenRequest;
+      case PermissionStatus.shouldRequest:
+        return PushNotificationPermissionStatus.shouldRequest;
+    }
+  }
+
+  @override
+  Future<int> getBadgeCount() {
+    if (!os.isIOS) {
+      _logger.error('Not supported on this platform');
+    }
+    if (!_isConfigured) {
+      throw _needsConfigurationException;
+    }
+    return _hostApi.getBadgeCount();
+  }
+
+  @override
+  Future<void> setBadgeCount(int badgeCount) async {
+    if (!os.isIOS) {
+      _logger.error('Not supported on this platform');
+    }
+    if (!_isConfigured) {
+      throw _needsConfigurationException;
+    }
+    await _hostApi.setBadgeCount(badgeCount);
+  }
+
+  Future<void> _registerBackgroundProcessorForAndroid() async {
+    final callbackHandle =
+        PluginUtilities.getCallbackHandle(_backgroundProcessor);
+    if (callbackHandle == null) {
+      throw const PushNotificationException(
+        'Callback is not a global or static function',
+        recoverySuggestion:
+            'Make the function a top-level or a static function.',
+      );
+    }
+    await _hostApi.registerCallbackFunction(
+      callbackHandle.toRawHandle(),
+    );
+    _logger.debug('Successfully registered callback');
+  }
+
+  Future<void> _registerDeviceWhenConfigure() async {
+    late String deviceToken;
+    try {
+      deviceToken = await _onTokenReceived.first;
+    } on PlatformException catch (error) {
+      // the error mostly like is the App doesn't have corresponding
+      // capability to request a push notification device token
+      throw PushNotificationException(
+        'Error occurred awaiting for device token',
+        recoverySuggestion: 'Please review the underlying exception',
+        underlyingException: error,
+      );
+    }
+    await _registerDevice(deviceToken);
+  }
+
+  void _foregroundNotificationListener(
+    PushNotificationMessage pushNotificationMessage,
+  ) =>
+      _serviceProviderClient.recordNotificationEvent(
+        eventType: PinpointEventType.foregroundMessageReceived,
+        notification: pushNotificationMessage,
+      );
+
+  void _notificationOpenedListener(
+    PushNotificationMessage pushNotificationMessage,
+  ) =>
+      _serviceProviderClient.recordNotificationEvent(
+        eventType: PinpointEventType.notificationOpened,
+        notification: pushNotificationMessage,
+      );
+
+  void _tokenReceivedListener(String deviceToken) {
+    unawaited(_registerDevice(deviceToken));
```

Comment:
Errors here should be added to the `onTokenReceived` stream, I think. Unawaiting will just report this as unhandled.

Reply:
This is just the internal listener and we are unawaiting since the registration process need not be blocking.